# STAT1100 Data Communication and Modelling

## Sorting Data and Identifying Duplicates

In [1]:
import pandas as pd
print(pd.__version__)

1.2.3


### Data sets

<a id='brainsize'></a>
#### Brain size

The `brainsize` data set is taken from Willerman, L., Schultz, R., Rutledge, J. N., and Bigler, E. (1991). In Vivo Brain Size and Intelligence, Intelligence, 15, 223-228.

The data are a sample of 40 right-handed university students (20 male and 20 female). Participants undertook the Wechsler Adult Intelligence Scale, resulting in three measurements: Full IQ, Verbal IQ, and Performance IQ. The researchers used Magnetic Resonance Imaging (MRI) to determine the brain size of the subjects (size was measured as total pixel count from the MRI scan). Information about gender and body size (height in inches and weight in pounds) are also included. The researchers withheld the weights of two subjects and the height of one subject for reasons of confidentiality.

In [2]:
brainsize = pd.read_csv("data/brainsize.csv")
brainsize.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40 entries, 0 to 39
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Gender     40 non-null     object 
 1   FSIQ       40 non-null     int64  
 2   VIQ        40 non-null     int64  
 3   PIQ        40 non-null     int64  
 4   Weight     38 non-null     float64
 5   Height     39 non-null     float64
 6   MRI_Count  40 non-null     int64  
dtypes: float64(2), int64(4), object(1)
memory usage: 2.3+ KB


There is also a "dirty" version of the `brainsize` data set that contains some duplicate and suspect rows for the purposes of illustrating the concepts here.

In [3]:
brainsize_dirty = pd.read_csv("data/brainsize_dirty.csv")
brainsize_dirty.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47 entries, 0 to 46
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Gender     47 non-null     object 
 1   FSIQ       47 non-null     int64  
 2   VIQ        47 non-null     int64  
 3   PIQ        46 non-null     float64
 4   Weight     42 non-null     float64
 5   Height     43 non-null     float64
 6   MRI_Count  47 non-null     int64  
dtypes: float64(3), int64(3), object(1)
memory usage: 2.7+ KB


<a id='CAERS'></a>
#### Adverse food

The `caers` data set is taken from a US FDA's Centre for Food Safety and Applied Nutrition (CFSAN) Adverse Events Reporting System (CAERS) report of reports of adverse events resulting from regulating products (including foods, dietary supplements, and cosmetics). The data description is shown below.

| Name of CAERS data field | Description |
|---|---|
| RA_Report | The unique number that identifies each case. |
| RA_CAERS Created Date | The date on which the data were first entered into CAERS from an adverse event report. |
| AEC_Event Start Date | The reported date on which the consumer first experienced the adverse event. |
| PRI_Product Role | Suspect or concomitant (as reported) |
| PRI_Reported Brand/Product Name | The verbatim brands and/or product names indicated to have been used by the consumer reported to have experienced the adverse event. An adverse event report may specify consumption of a single product or multiple products. |
| PRI_FDA Industry Code | The FDA industry code associated with the type of product reported. |
| PRI_FDA Industry Name | The FDA industry description associated with the type of product reported. (Ice cream products, cosmetics, Coffee/Tea) |
| CI_Age at Adverse Event | The age of the consumer reported to have experienced the adverse event. |
| CI_Age Unit | The time unit (day, week, month, year) of the age provided in the CI_Age at Adverse Event data field for the consumer reported to have experienced the adverse event. |
| CI_Gender | The sex of the individual reported to have experienced the adverse event. |
| AEC_One Row Outcomes | Outcome(s) of the adverse event experienced by the injured consumer as specified by the reporter; each report may indicate one or more outcomes for each consumer. |
| SYM_One Row Coded Symptoms | The symptom(s) experienced by the injured consumer as specified by the reporter and coded by FDA according to the Medical Data Dictionary for Regulatory Activities (MedDRA). Each adverse event report may indicate one or more symptoms for each consumer. |

In [4]:
caers = pd.read_csv("data/CAERS_ASCII_2004_2017Q2_modified.csv",
                    parse_dates=[1,2])
caers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90786 entries, 0 to 90785
Data columns (total 13 columns):
 #   Column                           Non-Null Count  Dtype         
---  ------                           --------------  -----         
 0   RA_Report                        90786 non-null  int64         
 1   RA_CAERS Created Date            90786 non-null  datetime64[ns]
 2   AEC_Event Start Date             53653 non-null  datetime64[ns]
 3   PRI_Product Role                 90786 non-null  object        
 4   PRI_Reported Brand/Product Name  90786 non-null  object        
 5   PRI_FDA Industry Code            90786 non-null  int64         
 6   PRI_FDA Industry Name            90786 non-null  object        
 7   CI_Age at Adverse Event          52926 non-null  float64       
 8   CI_Age Unit                      90786 non-null  object        
 9   CI_Gender                        90786 non-null  object        
 10  AEC_One Row Outcomes             90786 non-null  object   

### Sorting data

The [brainsize](#brainsize) data set is not sorted. The first five observations are shown below for comparison.

In [5]:
brainsize.head()

Gender  FSIQ  VIQ  PIQ  Weight  Height  MRI_Count
0  Female   133  132  124   118.0    64.5     816932
1    Male   140  150  124     NaN    72.5    1001121
2    Male   139  123  150   143.0    73.3    1038437
3    Male   133  129  128   172.0    68.8     965353
4  Female   137  132  134   147.0    65.0     951545

To arrange the observations in the order of the variable *FSIQ* we use the [sort_values()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html) method and the `by` parameter.

In [6]:
brainsize.sort_values(by="FSIQ").head()

Gender  FSIQ  VIQ  PIQ  Weight  Height  MRI_Count
28  Female    77   83   72   106.0    63.0     793549
19    Male    80   77   86   180.0    70.0     889083
38    Male    81   90   74   148.0    74.0     930016
34  Female    83   90   81   143.0    66.5     834344
20    Male    83   83   86     NaN     NaN     892420

To arrange the observations in descending order set the `ascending` parameter to `False`.

In [7]:
brainsize.sort_values(by="FSIQ", ascending=False).head()

Gender  FSIQ  VIQ  PIQ  Weight  Height  MRI_Count
31   Male   144  145  137   191.0    67.0     949589
25   Male   141  145  131   171.0    72.0     935494
11   Male   141  150  128   151.0    70.0    1079549
36   Male   140  150  124   144.0    70.5     949395
1    Male   140  150  124     NaN    72.5    1001121

Note that the sorting variable doesn't have to be a continuous variable, it can be a coded categorical or discrete valued variable, or even a variable containing strings which pandas will sort in alphanumeric order.

The following sort on the *Gender* variable which contains strings will list all the observations for female students first and then the observations for males since the first character 'F' precedes 'M'.

In [8]:
brainsize.sort_values(by="Gender").head()

Gender  FSIQ  VIQ  PIQ  Weight  Height  MRI_Count
0   Female   133  132  124   118.0    64.5     816932
22  Female   135  129  134   122.0    62.0     790619
28  Female    77   83   72   106.0    63.0     793549
29  Female   130  126  124   159.0    66.5     866662
18  Female   101  112   84   136.0    66.3     808020

In [9]:
brainsize.sort_values(by="Gender").tail()

Gender  FSIQ  VIQ  PIQ  Weight  Height  MRI_Count
3    Male   133  129  128   172.0    68.8     965353
2    Male   139  123  150   143.0    73.3    1038437
1    Male   140  150  124     NaN    72.5    1001121
25   Male   141  145  131   171.0    72.0     935494
39   Male    89   91   89   179.0    75.5     935863

We can sort a data frame on more than one variable and pandas gives priority to the order in which you specify the variables. Suppose we want to sort the data first by *FSIQ* and then by *MRI_Count*. Sorting on only *FSIQ* doesn't produce what we want. 

In [10]:
brainsize.sort_values(by="FSIQ").head(7)

Gender  FSIQ  VIQ  PIQ  Weight  Height  MRI_Count
28  Female    77   83   72   106.0    63.0     793549
19    Male    80   77   86   180.0    70.0     889083
38    Male    81   90   74   148.0    74.0     930016
34  Female    83   90   81   143.0    66.5     834344
20    Male    83   83   86     NaN     NaN     892420
15  Female    83   71   96   135.0    68.0     865363
26  Female    85   90   84   140.0    68.0     798612

The result of simply sorting on *FSIQ*, means that within a particular value of *FSIQ* the observations are not sorted in any particular order. Here, the three people with `FSIQ=83` have brain size 8.34, 8.92, and 8.65 hundred thousand pixels.

In [11]:
brainsize.sort_values(by=["FSIQ", "MRI_Count"]).head(7)

Gender  FSIQ  VIQ  PIQ  Weight  Height  MRI_Count
28  Female    77   83   72   106.0    63.0     793549
19    Male    80   77   86   180.0    70.0     889083
38    Male    81   90   74   148.0    74.0     930016
34  Female    83   90   81   143.0    66.5     834344
15  Female    83   71   96   135.0    68.0     865363
20    Male    83   83   86     NaN     NaN     892420
26  Female    85   90   84   140.0    68.0     798612

The order that multiple sorting variables are specified is very important! To see this, try sorting the data with the variables *MRI_Count* and *FSIQ* swapped.

In [12]:
brainsize.sort_values(by=["MRI_Count", "FSIQ"]).head()

Gender  FSIQ  VIQ  PIQ  Weight  Height  MRI_Count
22  Female   135  129  134   122.0    62.0     790619
28  Female    77   83   72   106.0    63.0     793549
26  Female    85   90   84   140.0    68.0     798612
18  Female   101  112   84   136.0    66.3     808020
0   Female   133  132  124   118.0    64.5     816932

In this case, it was redundant to sort by *FSIQ* after sorting by *MRI_Count* because all of the brain sizes were unique in this data set. So there were no groups of common *MRI_Count* values to sort within.

When you use the [sort_values()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html) method in this way, it returns a copy of the data frame that is sorted without affecting or overwriting the original data frame. To change the order of the observations in the original data frame, set the parameter `inplace=True`.

### Checking for duplicate rows

Below are the first ten rows of the "dirty" [brainsize](#brainsize) data set which has 47 observations.

In [13]:
brainsize_dirty.head(10)

Gender  FSIQ  VIQ   PIQ  Weight  Height  MRI_Count
0  Female    77   83  72.0   106.0    63.0     793549
1  Female    77   83  72.0   106.0    63.0     893549
2    Male    80   77  86.0   180.0    70.0     889083
3    Male    81   90  74.0   148.0    74.0     930016
4  Female    83   71  96.0   135.0    68.0     865363
5    Male    83   83  86.0     NaN     NaN     892420
6    male    83   83  86.0     NaN     NaN     892420
7    Male    83   83  86.0     NaN     NaN     892420
8    Male    83   83   NaN     NaN     NaN     892420
9  Female    83   90  81.0   143.0    66.5     834344

Observations 0 and 1 are not duplicates because although they are similar on most variables, the *MRI_Count* is not the same (the leading digit differs). If this were real data that matched so closely, we might suspect a data entry error and that these rows were supposed to pertain to the same individual.

Observations 5 and 7 are identical on all variables. Observation 6 has the gender in lower case so it won't be counted as a duplicate. Observation 8 matches on all non-missing values but has missing data for the *PIQ* variable so it will also not be identified as a duplicate row. If this were the real-life state of the data, we would suspect that these four rows are likely to belong to the same person (since it's very unlikely for two people to have exactly the same 6 digit number for brain size in pixels). But the only true duplicate from a data perspective is observation 7.

Let's use the [duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) method to show and count the number of duplicate rows and then use the [drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) method to remove the duplicate rows and save the resulting data to a new data frame `brainsize_nodups`. All columns are used for identifying duplicates by default.

In [14]:
brainsize_dirty.duplicated().head(10)

0    False
1    False
2    False
3    False
4    False
5    False
6    False
7     True
8    False
9    False
dtype: bool

In [15]:
brainsize_dirty[brainsize_dirty.duplicated()]

Gender  FSIQ  VIQ    PIQ  Weight  Height  MRI_Count
7     Male    83   83   86.0     NaN     NaN     892420
31    Male   133  129  128.0   172.0    68.8     965353
33  Female   133  132  124.0   118.0    64.5     816932

In [16]:
brainsize_dirty.duplicated().sum()

3

In [17]:
(~brainsize_dirty.duplicated()).sum()

44

In [18]:
brainsize_nodups = brainsize_dirty.drop_duplicates()
brainsize_nodups.shape

(44, 7)

In [19]:
brainsize_nodups.head(10)

Gender  FSIQ  VIQ   PIQ  Weight  Height  MRI_Count
0   Female    77   83  72.0   106.0    63.0     793549
1   Female    77   83  72.0   106.0    63.0     893549
2     Male    80   77  86.0   180.0    70.0     889083
3     Male    81   90  74.0   148.0    74.0     930016
4   Female    83   71  96.0   135.0    68.0     865363
5     Male    83   83  86.0     NaN     NaN     892420
6     male    83   83  86.0     NaN     NaN     892420
8     Male    83   83   NaN     NaN     NaN     892420
9   Female    83   90  81.0   143.0    66.5     834344
10  Female    85   90  84.0   140.0    68.0     798612

The data set has been reduced from 47 observations to 44 observations and there are only three non-duplicate observations for the person with `MRI_Count=892420`.

In [20]:
brainsize_nodups[brainsize_nodups["MRI_Count"]==816932]

Gender  FSIQ  VIQ    PIQ  Weight  Height  MRI_Count
32   Female   133  132  124.0   118.0    64.5     816932
34  Female    133  132  124.0   118.0    64.5     816932

Why are the two observations for the person with `MRI_Count=816932` not considered duplicates? Investigate the data.

### Checking for duplicate variable values

The [caers](#CAERS) data set has several instances of rows with duplicate report numbers. Let's pick one of them `RA_Report=65420`.

In [21]:
caers[caers["RA_Report"]==65420].iloc[:,[0,4,7,9,10]]

RA_Report PRI_Reported Brand/Product Name  CI_Age at Adverse Event  \
22      65420                          COFFEE                     33.0   
23      65420         DIET CAFFEINE FREE COKE                     33.0   
24      65420                       DIET COKE                     33.0   
25      65420                      DIET PEPSI                     33.0   
26      65420                     DIET SPRITE                     33.0   
27      65420                            JAMS                     33.0   
28      65420                LISTERINE STRIPS                     33.0   
29      65420                      NUTRASWEET                     33.0   
30      65420                        PUDDINGS                     33.0   
31      65420                             TEA                     33.0   

   CI_Gender          AEC_One Row Outcomes  
22      Male  LIFE THREATENING, DISABILITY  
23      Male  LIFE THREATENING, DISABILITY  
24      Male  LIFE THREATENING, DISABILITY  
25      Male  LIFE THREATENING, DISABILITY  
26      Male  LIFE THREATENING, DISABILITY  
27      Male  LIFE THREATENING, DISABILITY  
28      Male  LIFE THREATENING, DISABILITY  
29      Male  LIFE THREATENING, DISABILITY  
30      Male  LIFE THREATENING, DISABILITY  
31      Male  LIFE THREATENING, DISABILITY

It appears that these rows pertain to a single report of a 33 year old male who had a life threatening adverse reaction to one or more of 10 suspect products. Hence this data set does not contain records of unique people, there is one row for every product listed in each report.

Suppose we wanted to know how many adverse events there were for males and females. What is wrong with simply tabulating the *CI_Gender* variable?

In [22]:
caers["CI_Gender"].value_counts()

Female           58924
Male             26943
Not Available     4916
Unknown              2
Not Reported         1
Name: CI_Gender, dtype: int64

In [23]:
caers.shape

(90786, 13)

There are 90,786 rows in the data set. But the problem is that the 33 year old male above will have been counted 10 times for the same report (same adverse event) in this frequency table. To identify how many unique reports there were from males and females we can use the [duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) and [drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) methods, and only consider the *RA_Report* column for identifying duplicates.

In [24]:
caers.duplicated(["RA_Report"]).sum()

26269

In [25]:
caers_unique = caers.drop_duplicates(["RA_Report"])

In [26]:
caers_unique.shape

(64517, 13)

The data set has been reduced to 64,517 observations after 26,269 rows with duplicate report numbers were deleted. The data set now contains one row per report. We can assume that this is approximately the same as the number of individual consumers who had an adverse event, but it is possible for someone to have had two events and thus two report numbers.

In [27]:
caers_unique["CI_Gender"].value_counts()

Female           40815
Male             19655
Not Available     4044
Unknown              2
Not Reported         1
Name: CI_Gender, dtype: int64

So the total number of report numbers in the data set is 64,517, of which 40,815 pertained to females and 19,655 pertained to males.

In [28]:
caers_unique["CI_Gender"].value_counts()/caers_unique.shape[0]

Female           0.632624
Male             0.304648
Not Available    0.062681
Unknown          0.000031
Not Reported     0.000015
Name: CI_Gender, dtype: float64

From this table we can conclude that about 63% of reports were adverse events in female consumers, and 30% of reports were adverse events in male consumers (with the remaining reports missing gender data).